hnix-store-core/src/System/Nix/Internal/Hash.hs

```diff
@@ -0,0 +1,152 @@
+import qualified Data.ByteString.Lazy  as BSL
+import qualified Data.Hashable         as DataHashable
+import           Data.List             (foldl')
+import           Data.Proxy            (Proxy(Proxy))
+import qualified Data.Text             as T
+import qualified Data.Text.Encoding    as T
+import qualified Data.Vector           as V
+import           Data.Word             (Word8)
+
+-- | A tag for different hashing algorithms
+--   Also used as a type-level tag for hash digests
+--   (e.g. @Digest SHA256@ is the type for a sha256 hash)
+data HashAlgorithm
+  = MD5
+  | SHA1
+  | SHA256
+
+
+-- | Types with kind @HashAlgorithm@ may be a @HasDigest@ instance
+--   if they are able to hash bytestrings via the init/update/finalize
+--   API of cryptonite
+--
+--   Each instance defined here simply defers to one of the underlying
+--   monomorphic hashing libraries, such as `cryptohash-sha256`.
+class HasDigest (a :: HashAlgorithm) where
+
+  type AlgoCtx a :: *
+  hashLength        :: Proxy a -> Int
+
+  validateHashBytes :: BS.ByteString -> Maybe (Digest a)
+  validateHashBytes bs = if BS.length bs == hashLength (Proxy @a)
+                         then Just (Digest bs)
+                         else Nothing
+
+  initialize        :: AlgoCtx a
+  update            :: AlgoCtx a -> BS.ByteString -> AlgoCtx a
+  finalize          :: AlgoCtx a -> Digest a
+
+  toBase32          :: Digest a -> T.Text
+  -- ^ The base32 representation of a @Digest a@ is a text
+  --   encoding. TODO - reference thesis spec?
+
+
+-- | The cryptographic hash of of a strict bytestring, where hash
+--   algorithm is chosen by the type of the digest
+--   For example:
+--   > let d = hash "Hello, sha-256!" :: Digest SHA256
+--   or
+--   > :set -XTypeApplications
+--   > let d = hash @SHA256 "Hello, sha-256!"
+hash :: forall a.HasDigest a => BS.ByteString -> Digest a
+hash bs =
+  finalize $ update @a (initialize @a) bs
+
+-- | The cryptographic hash of a lazy bytestring. Use is the same
+--   as for @hash@. This runs in constant space, but forces the
+--   entire bytestring
+hashLazy :: forall a.HasDigest a => BSL.ByteString -> Digest a
+hashLazy bsl =
+  finalize $ foldl' (update @a) (initialize @a) (BSL.toChunks bsl)
+
+
+instance HasDigest MD5 where
+  type AlgoCtx 'MD5 = MD5.Ctx
+  hashLength _ = 16
+  initialize = MD5.init
+  update = MD5.update
+  finalize = Digest . MD5.finalize
+  toBase32 (Digest bs) = T.decodeUtf8 $ printHashBytes32 bs
+
+instance HasDigest 'SHA1 where
+  type AlgoCtx SHA1 = SHA1.Ctx
+  hashLength _ = 20
+  initialize = SHA1.init
+  update = SHA1.update
+  finalize = Digest . SHA1.finalize
+  toBase32 (Digest bs) = T.decodeUtf8 $ printHashBytes32 bs
+
+instance HasDigest 'SHA256 where
+  type AlgoCtx SHA256 = SHA256.Ctx
+  hashLength _ = 32
+  initialize = SHA256.init
+  update = SHA256.update
+  finalize = Digest . SHA256.finalize
+  toBase32 (Digest bs) =
+    T.decodeUtf8 $ printHashBytes32 $ truncateDigest 20 bs
```

Comment:
? We definitely need untruncated sha-256 hashes (and their base 32 representations) as well as truncated ones, and in any case if we're dealing with truncated hashes it's the entire hash that's truncated, not just its human-friendly representation (e.g. when comparing for equality). IMO `HashAlgorithm` should have a `Truncated` constructor taking another `HashAlgorithm`

Reply:
```
let h_bytes = hash256(string)
    h_bytes' = truncate(h_bytes)
    forStorePath = toBase32(h_bytes')
```
Imagining I go through these steps to produce a store path hash. There are multiple places in the code where we need the untruncated digest, and multiple places where we need the final base32 encoded truncated hash.

Is there any place in the code other than generating base32 encodings, where we would need access to the truncated version of the hash bytes? If so, then a `Truncated` newtype would be useful for carrying the proof that truncation has happened. OTOH if truncation is merely an implementation detail of producing base32 encoding of a hash, then `Truncated` newtype wouldn't pull its weight. What do you think?